timers/timers.go

```diff
@@ -0,0 +1,164 @@
+package timers
+
+import (
+	"runtime"
+	"sync"
+	"time"
+	"unsafe"
+
+	"github.com/uber-go/atomic"
+)
+
+const (
+	cacheline = 64
+
+	// State machine for timers.
+	stateDefault = iota
+	stateExpired
+	stateCanceled
+)
+
+type mutex struct {
+	sync.Mutex
+	// Avoid false sharing:
+	// http://mechanical-sympathy.blogspot.com/2011/07/false-sharing.html
+	pad [cacheline - unsafe.Sizeof(sync.Mutex{})]byte
+}
+
+// Timer FIXME
+type Timer struct {
+	f        func()
+	deadline uint64 // in ticks
+	state    atomic.Int32
+	prev     *Timer
+}
+
+// Stop FIXME
+func (t *Timer) Stop() {
+	t.state.Store(stateCanceled)
+}
+
+func (t *Timer) append(tail *Timer) *Timer {
+	tail.prev = t
+	return tail
+}
+
+// Wheel FIXME
+type Wheel struct {
```

Comment:
I think we could simplify this further for our use case:
 - we know what the maximum timeout is (2 mins), and we can reject timers set past this timeout.
 - if we're thinking of starting with 5ms buckets, that means we'll only have 400 buckets
 - since this is a pretty small number, having a lock per bucket seems reasonable
 - whenever we need to add a timer, we have to bring the bucket into the cache line anyway, so having the mutex + bucket on the same cache line makes sense

So I think we can get rid of the bucket mask and lock mask, and possibly also the padded mutex, since we could do something like:
```go
type bucket struct {
  sync.Mutex
  timers []*Timer
  timersBacking [1000]*Timer
}
```

The `timersBacking` would ensure no sharing by making `bucket` (way) too large to fit into a cache line, and means the initial timers are right next to the mutex.

In terms of memory usage, each bucket is 4 pointers + size of `timersBacking` * pointers per bucket. If we go with 1000 timers, that's ~8k per bucket. 8k * 400 buckets = ~3.2MB, which seems pretty reasonable since we'll only have a single wheel per channel.

Since we have plenty of RAM, we could probably bump up `timersBacking` if needed (e.g., if we get ~50k requests per second in a channel, we'd probably want 50k/5ms = 10000 which is still reasonable.)